File: packages/swift-sdk/SwiftExampleApp/SwiftExampleApp/Core/Services/PlatformBalanceSyncService.swift
// PlatformBalanceSyncService.swift
// SwiftExampleApp
//
// App-level service that reflects Rust-owned BLAST address sync state
// into SwiftUI. Scheduling, incremental state, and balance tracking
// are handled on the Rust side.

import Foundation
import SwiftUI
import Combine
import SwiftData
import SwiftDashSDK

/// Observable service managing BLAST address balance sync UI state.
///
/// The Rust manager owns the background sync loop; this service loads
/// cached balances for immediate display, issues manual sync requests,
/// and applies Rust-emitted completion events to the UI.
@MainActor
class PlatformBalanceSyncService: ObservableObject {
    // MARK: - Published State

    /// Whether a sync is currently in progress.
    @Published var isSyncing = false

    /// Whether a `clearLocalState` is currently in progress. Gates the
    /// Clear / Sync Now controls so a second Clear or a Sync Now can't
    /// interleave with the Rust reset + SwiftData wipe sequence.
    @Published var isClearing = false

    /// Monotonic token bumped at the start of every `clearLocalState`. A
    /// `refreshBalanceSnapshot` captures it before its `await` and
    /// re-checks it before publishing, so a snapshot that was already in
    /// flight when a Clear ran can't republish pre-clear balances over
    /// the cleared UI. `@MainActor` serializes the bump and the check.
    private var balanceSnapshotGeneration: UInt64 = 0

    /// Last successful sync time (local clock).
    @Published var lastSyncTime: Date?

    /// Per-address balances (address hash hex → balance).
    @Published var addressBalances: [String: UInt64] = [:]

    /// Aggregate platform balance across all synced addresses.
    @Published var totalPlatformBalance: UInt64 = 0

    /// Number of addresses with non-zero balance.
    @Published var activeAddressCount: Int = 0

    /// Checkpoint height from the last trunk/branch tree scan.
    @Published var checkpointHeight: UInt64 = 0

    /// Chain tip height (highest block seen from incremental catch-up).
    @Published var chainTipHeight: UInt64 = 0

    /// Last sync height for display.
    @Published var lastSyncHeight: UInt64 = 0

    /// Last block height with recent address changes (for compaction detection).
    @Published var lastKnownRecentBlock: UInt64 = 0

    /// Platform block time from the most recent sync (Unix seconds).
    @Published var lastSyncBlockTime: Date?

    /// Total number of successful syncs since launch.
    @Published var syncCountSinceLaunch: Int = 0

    /// Cumulative query counts since launch.
    @Published var totalTrunkQueries: UInt32 = 0
    @Published var totalBranchQueries: UInt32 = 0
    @Published var totalCompactedQueries: UInt32 = 0
    @Published var totalRecentQueries: UInt32 = 0
    @Published var totalRecentEntries: UInt32 = 0
    @Published var totalCompactedEntries: UInt32 = 0

    /// Raw GroveDB proof from the most recent sync (for debugging).
    @Published var lastRecentProof: Data = Data()

    /// Last error message, cleared on successful sync.
    @Published var lastError: String?

    // MARK: - Internal

    /// The platform address wallet handle used to refresh address
    /// balances after Rust-owned sync passes complete.
    private var platformAddressWallet: ManagedPlatformAddressWallet?

    /// Wallet manager used to control the Rust-owned background sync loop.
    private weak var walletManager: PlatformWalletManager?

    /// Persistence handler for loading cached balances.
    private var persistenceHandler: PlatformWalletPersistenceHandler?

    /// Wallet ID for querying cached balances.
    private var walletId: Data?

    /// Observes Rust-side sync completion events.
    private var syncEventCancellable: AnyCancellable?

    /// Mirrors the Rust manager's current `is_syncing` flag for the UI.
    private var syncStateCancellable: AnyCancellable?

    // MARK: - Lifecycle

    /// Configure for a wallet. Call after wallet creation/switch.
    func configure(
        platformAddressWallet: ManagedPlatformAddressWallet,
        walletManager: PlatformWalletManager,
        persistenceHandler: PlatformWalletPersistenceHandler? = nil,
        walletId: Data? = nil
    ) {
        // Same invalidation as `reset()`: a snapshot still in flight for
        // the previous wallet must not publish over this configuration.
        balanceSnapshotGeneration &+= 1
        self.platformAddressWallet = platformAddressWallet
        self.walletManager = walletManager
        self.persistenceHandler = persistenceHandler
        self.walletId = walletId
        self.syncEventCancellable?.cancel()
        self.syncStateCancellable?.cancel()

        // Load cached state from SwiftData for immediate display.
        if let handler = persistenceHandler, let wid = walletId {
            // Restore address balances for UI.
            let cached = handler.loadCachedBalances(walletId: wid)
            if !cached.isEmpty {
                var newBalances: [String: UInt64] = [:]
                var total: UInt64 = 0
                var nonZero = 0

                for (_, hash, balance, _, _, _, _) in cached {
                    let key = hash.map { String(format: "%02x", $0) }.joined()
                    newBalances[key] = balance
                    total += balance
                    if balance > 0 { nonZero += 1 }
                }

                addressBalances = newBalances
                totalPlatformBalance = total
                activeAddressCount = nonZero
            }

            // Load the last persisted network sync markers for display.
            if let state = handler.loadCachedSyncState(walletId: wid) {
                chainTipHeight = state.syncHeight
                lastSyncHeight = state.syncHeight
                if state.syncTimestamp > 0 {
                    lastSyncBlockTime = Date(timeIntervalSince1970: TimeInterval(state.syncTimestamp))
                }
                lastKnownRecentBlock = state.lastKnownRecentBlock

                SDKLogger.log(
                    "Restored network sync state: height=\(state.syncHeight), timestamp=\(state.syncTimestamp)",
                    minimumLevel: .medium
                )
            }
        }

        syncStateCancellable = walletManager.$platformAddressSyncIsSyncing
            .sink { [weak self] isSyncing in
                self?.isSyncing = isSyncing
            }

        syncEventCancellable = walletManager.$lastPlatformAddressSyncEvent
            .sink { [weak self] event in
                guard let self, let event else { return }
                self.handlePlatformAddressSyncEvent(event)
            }
    }

    /// Clear UI display state — balances, metrics, last sync timestamps.
    ///
    /// Does NOT nil out the wallet handle, so the user can tap "Sync Now"
    /// immediately after clearing. Use [`fullReset`] for wallet deletion
    /// or network switches.
    func clearDisplay() {
        addressBalances.removeAll()
        totalPlatformBalance = 0
        activeAddressCount = 0
        checkpointHeight = 0
        chainTipHeight = 0
        lastSyncHeight = 0
        lastKnownRecentBlock = 0
        lastSyncBlockTime = nil
        lastRecentProof = Data()
        lastError = nil
        lastSyncTime = nil
        syncCountSinceLaunch = 0
        totalTrunkQueries = 0
        totalBranchQueries = 0
        totalCompactedQueries = 0
        totalRecentQueries = 0
        totalRecentEntries = 0
        totalCompactedEntries = 0
    }

    /// Full reset — clears display state AND nils out the wallet handle.
    /// Use for wallet deletion or network switch. Caller must re-configure
    /// before the next sync.
    func reset() {
        // Invalidate any in-flight `refreshBalanceSnapshot` — its
        // detached task captured the OLD wallet handle strongly, so
        // without this bump a snapshot racing a wallet/network switch
        // would publish the old wallet's balances over the reset UI.
        balanceSnapshotGeneration &+= 1
        clearDisplay()
        platformAddressWallet = nil
        walletManager = nil
        syncEventCancellable?.cancel()
        syncStateCancellable?.cancel()
    }

    /// Clear platform-address sync data for real — what the Platform
    /// Sync "Clear" button calls.
    ///
    /// Plain [`clearDisplay`] only zeroes the in-memory `@Published`
    /// mirror, so the next sync resumed from the surviving watermark in
    /// ~2s (the "Clear didn't work" symptom). This clears the stores the
    /// synced data actually lives in, mirroring
    /// `ShieldedService.clearLocalState`: Rust-side reset FIRST (so the
    /// next sync can't re-persist stale rows), then the SwiftData clear,
    /// then the published-mirror reset.
    ///
    /// The `PersistentPlatformAddress` rows are **zeroed in place, not
    /// deleted** — they carry durable derivation metadata that no sync
    /// re-emits in-session and that the balance callback only updates
    /// (never creates), so deleting them would leave the UI empty until
    /// app restart. Only the volatile balance/sync fields are cleared;
    /// the network-keyed `PersistentPlatformAddressesSyncState` watermark
    /// row is deleted to force a full rescan.
    ///
    /// Scoped to the **active network**. The SwiftData store holds rows
    /// for every network at once (the UI filters them by network), so
    /// clearing must not touch other networks' state. `PersistentPlatformAddress`
    /// carries no network column, so it's scoped via `walletIdsOnNetwork`
    /// — the same wallet-id-per-network pivot the view uses;
    /// `PersistentPlatformAddressesSyncState` is network-keyed and is
    /// scoped by `networkRaw`. This matches the manager-level Rust reset,
    /// which only touches the active network's registered wallets.
    ///
    /// Fails closed: if the Rust reset OR the SwiftData delete throws, it
    /// surfaces the error in `lastError` and returns WITHOUT calling
    /// `clearDisplay()` — the UI never shows a false "cleared" state over
    /// data that is still on disk / in Rust memory.
    func clearLocalState(
        modelContext: ModelContext,
        network: Network,
        walletIdsOnNetwork: Set<Data>
    ) async {
        // Invalidate any in-flight `refreshBalanceSnapshot`: a snapshot
        // queued by a sync that completed just before this Clear must not
        // land its (pre-clear) balances over the cleared UI. It captured
        // the old generation and will drop its publish once we bump here.
        balanceSnapshotGeneration &+= 1
        // Gate the Clear / Sync Now controls for the whole sequence.
        isClearing = true
        defer { isClearing = false }

        // 1) Reset the Rust-owned state BEFORE touching disk. Without
        //    this the in-memory watermark survives and the next "Sync
        //    Now" resumes incrementally (fast) instead of doing a full
        //    rescan; a still-registered background pass could also
        //    re-persist the rows we're about to delete. Fail closed —
        //    the reset is load-bearing for the wipe, so abort (surfacing
        //    the error) rather than leave a half-cleared state.
        if let walletManager {
            do {
                try await walletManager.resetPlatformAddressSyncState()
            } catch {
                lastError = "Failed to reset platform-address sync state: \(error.localizedDescription)"
                SDKLogger.error(lastError ?? "")
                return
            }
        }

        // 2) Zero the volatile balance/sync fields of this network's
        //    address rows IN PLACE — do NOT delete them. A
        //    `PersistentPlatformAddress` row is durable derivation state
        //    (bech32m address, 20-byte hash, 33-byte public key,
        //    account/address index, derivation path), not just a balance
        //    cache. The BLAST balance callback `persistAddressBalances`
        //    only *updates* existing rows (it skips by `addressHash` when
        //    the row is missing), and no sync re-emits the rows in-session:
        //    Rust's `reset_sync_state` preserves the address bijection, so
        //    there is no pool extension to re-fire the address-emit path.
        //    Deleting would therefore leave the next "Sync Now" with
        //    nothing to upsert against — balances (and the spend/signing
        //    derivation lookups) would stay gone until an app restart
        //    re-seeds the rows. Zeroing clears what the user sees while
        //    keeping the metadata the rescan needs.
        do {
            let addresses = try modelContext.fetch(FetchDescriptor<PersistentPlatformAddress>())
            for row in addresses where walletIdsOnNetwork.contains(row.walletId) {
                row.balance = 0
                row.nonce = 0
                row.isUsed = false
                row.firstSeenHeight = 0
                row.lastSeenHeight = 0
                row.lastUpdated = Date()
            }

            // The sync-state watermark is pure volatile sync state (no
            // durable metadata) and is what actually forces a full rescan,
            // so delete it outright — the next sync re-creates it.
            let networkRaw = network.rawValue
            let syncStates = try modelContext.fetch(FetchDescriptor<PersistentPlatformAddressesSyncState>())
            for row in syncStates where row.networkRaw == networkRaw {
                modelContext.delete(row)
            }

            try modelContext.save()
        } catch {
            lastError = "Failed to clear persisted platform-address state: \(error.localizedDescription)"
            SDKLogger.error(lastError ?? "")
            return
        }

        // 3) Zero the published display mirror — only on full success.
        clearDisplay()
    }

    /// Trigger a manual sync. No-op if already syncing.
    func manualSync() async {
        await performSync()
    }

    // MARK: - Sync

    /// Perform the actual BLAST address sync via platform-wallet.
    ///
    /// Drives `isSyncing` directly around the await so the spinner
    /// flashes even when the underlying Rust pass completes faster
    /// than the manager's 1 Hz `isPlatformAddressSyncing` poll
    /// cadence. Empty-pool regtest syncs return in tens of ms — the
    /// polled `$platformAddressSyncIsSyncing` stays `false` for the
    /// whole call and the subscription never fires, so we have to
    /// reset locally regardless of outcome.
    func performSync() async {
        guard !isSyncing else { return }
        // The Clear/Sync mutual exclusion must live here, not only on
        // CoreContentView's buttons: pull-to-refresh (Wallets/Identities
        // tabs) and the post-submit resyncs (Transfer/Withdraw views)
        // call performSync directly and would otherwise start a sync
        // between clearLocalState's Rust reset and its SwiftData wipe.
        guard !isClearing else { return }
        guard let walletManager = walletManager else {
            lastError = "Platform address wallet not configured"
            return
        }

        isSyncing = true
        lastError = nil
        defer { isSyncing = false }

        do {
            try await walletManager.syncPlatformAddressNow()
        } catch {
            lastError = error.localizedDescription
            SDKLogger.log(
                "BLAST sync error: \(error.localizedDescription)",
                minimumLevel: .medium
            )
        }
    }

    private func handlePlatformAddressSyncEvent(_ event: PlatformAddressSyncEvent) {
        guard let walletId, let result = event.result(for: walletId) else {
            return
        }

        if result.success {
            lastError = nil
            if result.checkpointHeight > 0 {
                checkpointHeight = result.checkpointHeight
            }
            if result.newSyncHeight > chainTipHeight {
                chainTipHeight = result.newSyncHeight
            }
            lastSyncHeight = result.newSyncHeight
            lastKnownRecentBlock = result.lastKnownRecentBlock
            if result.newSyncTimestamp > 0 {
                lastSyncBlockTime = Date(timeIntervalSince1970: TimeInterval(result.newSyncTimestamp))
            }

            totalTrunkQueries += result.metrics.trunkQueries
            totalBranchQueries += result.metrics.branchQueries
            totalCompactedQueries += result.metrics.compactedQueries
            totalRecentQueries += result.metrics.recentQueries
            totalRecentEntries += result.metrics.recentEntriesReturned
            totalCompactedEntries += result.metrics.compactedEntriesReturned

            lastSyncTime = Date(timeIntervalSince1970: TimeInterval(event.syncUnixSeconds))
            syncCountSinceLaunch += 1

            Task {
                await refreshBalanceSnapshot()
            }
        } else {
            lastError = result.errorMessage ?? "Platform address sync failed"
        }
    }

    private func refreshBalanceSnapshot() async {
        guard let wallet = platformAddressWallet else { return }

        // Capture the generation before the off-actor read. If a
        // `clearLocalState` bumps it while we await, the balances below
        // are pre-clear and must NOT be published over the cleared UI.
        let generation = balanceSnapshotGeneration

        do {
            let (balances, credits) = try await Task.detached { [wallet] in
                let balances = try wallet.addressesWithBalances()
                let credits = try wallet.totalCredits()
                return (balances, credits)
            }.value

            // A Clear ran while this snapshot was in flight — drop it.
            guard generation == balanceSnapshotGeneration else {
                SDKLogger.log(
                    "BLAST snapshot dropped: superseded by a Clear",
                    minimumLevel: .medium
                )
                return
            }

            var newBalances: [String: UInt64] = [:]
            var total: UInt64 = 0
            var nonZero = 0

            for entry in balances {
                let key = entry.hash.map { String(format: "%02x", $0) }.joined()
                newBalances[key] = entry.balance
                total += entry.balance
                if entry.balance > 0 { nonZero += 1 }
            }

            addressBalances = newBalances
            totalPlatformBalance = credits == total ? total : credits
            activeAddressCount = nonZero

            SDKLogger.log(
                "BLAST sync complete: \(balances.count) addresses, total balance: \(totalPlatformBalance)",
                minimumLevel: .medium
            )
        } catch {
            lastError = error.localizedDescription
            SDKLogger.log(
                "BLAST sync snapshot refresh error: \(error.localizedDescription)",
                minimumLevel: .medium
            )
        }
    }
}

File: packages/kotlin-sdk/sdk/src/main/kotlin/org/dashfoundation/dashsdk/services/PlatformBalanceSyncService.kt
package org.dashfoundation.dashsdk.services

import androidx.room.withTransaction
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import org.dashfoundation.dashsdk.persistence.DashDatabase
import org.dashfoundation.dashsdk.persistence.syncStateScopeId
import org.dashfoundation.dashsdk.wallet.PlatformWalletManager
import org.dashfoundation.dashsdk.wallet.WalletSyncEvent

/**
 * Reactive view of the Rust-owned platform-address (BLAST) sync loop —
 * port of `PlatformBalanceSyncService.swift`.
 *
 * The Swift service spreads its state across `@Published` scalars; here it
 * is a single sealed [PlatformSyncState] StateFlow (idle / syncing /
 * synced(height, timestamp) / error) fed by two sources exactly as Swift
 * does:
 *
 * 1. **Completion events** — [PlatformWalletManager.syncEvents] carries the
 *    per-wallet [WalletSyncEvent.PlatformAddressResult] and the
 *    [WalletSyncEvent.PlatformAddressPassCompleted] boundary from the native
 *    `on_platform_address_sync_completed_fn` callback.
 * 2. **Watermark rows** — the network-scoped
 *    `PlatformAddressesSyncStateEntity` row Rust's persister writes on each
 *    pass, read on [configure] for the startup snapshot (Swift's
 *    `loadCachedSyncState`).
 *
 * [totalPlatformBalance] / [activeAddressCount] are Room `Flow`s straight
 * off [org.dashfoundation.dashsdk.persistence.dao.PlatformAddressDao], the
 * analogue of Swift's `refreshBalanceSnapshot()` sums.
 *
 * This service REFLECTS the loop; it never starts/stops it (that stays on
 * [PlatformWalletManager], which wraps the Rust start/stop entry points).
 * The one exception is [manualSync], which forwards to the manager's
 * user-initiated `sync_now` wrapper — a request, not orchestration.
 */
class PlatformBalanceSyncService(private val database: DashDatabase) {

    /** Sealed reduction of the platform-address sync loop's state. */
    sealed interface PlatformSyncState {
        /** No pass has been observed and none is in flight. */
        data object Idle : PlatformSyncState

        /** A pass is running (mirrors `walletManager.platformAddressSyncIsSyncing`). */
        data object Syncing : PlatformSyncState

        /**
         * A pass completed successfully.
         *
         * @param syncHeight the incremental-sync watermark height.
         * @param syncTimestamp unix seconds of the watermark block (0 if unknown).
         */
        data class Synced(val syncHeight: Long, val syncTimestamp: Long) : PlatformSyncState

        /** The last pass reported a failure for at least one wallet. */
        data class Error(val message: String) : PlatformSyncState
    }

    private val scope = CoroutineScope(SupervisorJob())

    private val _state = MutableStateFlow<PlatformSyncState>(PlatformSyncState.Idle)

    /** Current reduced sync state. */
    val state: StateFlow<PlatformSyncState> = _state.asStateFlow()

    /**
     * Total platform-credit balance across all addresses (sum of non-zero
     * balances) — Swift's `totalPlatformBalance`. Live Room Flow.
     */
    val totalPlatformBalance: Flow<Long> =
        database.platformAddressDao().observeNonZeroBalances()
            .map { rows -> rows.sumOf { it.balance } }

    /** Count of addresses carrying a non-zero balance — Swift's `activeAddressCount`. */
    val activeAddressCount: Flow<Int> =
        database.platformAddressDao().observeNonZeroBalances()
            .map { it.size }

    private var manager: PlatformWalletManager? = null
    private var eventJob: Job? = null
    private var pollJob: Job? = null

    /**
     * Bind to [manager] and start reflecting its sync loop — port of Swift
     * `configure(...)`. Loads the cached watermark for the manager's network
     * (startup snapshot), then subscribes to completion events and a
     * liveness poll fallback. Re-configuring rebinds cleanly.
     */
    fun configure(manager: PlatformWalletManager) {
        reset()
        this.manager = manager

        // Startup snapshot from the persisted watermark (Swift's
        // loadCachedSyncState): a prior successful pass shows as Synced.
        scope.launch {
            val scopeId = syncStateScopeId(manager.network.ffiValue)
            database.platformAddressDao().getSyncState(scopeId)?.let { row ->
                if (_state.value is PlatformSyncState.Idle) {
                    _state.value = PlatformSyncState.Synced(row.syncHeight, row.syncTimestamp)
                }
            }
        }

        // Reduce completion events into the sealed state.
        eventJob = scope.launch {
            manager.syncEvents.collect { event -> reduce(event) }
        }

        // Liveness poll fallback at the manager's cadence — flips to Syncing
        // when a pass is running even if the completion event is missed.
        pollJob = scope.launch {
            while (true) {
                val running = runCatching { manager.isPlatformAddressSyncRunning() }
                    .getOrDefault(false)
                if (running && _state.value !is PlatformSyncState.Syncing) {
                    // Only enter Syncing from a settled state; don't clobber a
                    // freshly-published Synced/Error from the event path.
                    if (_state.value is PlatformSyncState.Idle) {
                        _state.value = PlatformSyncState.Syncing
                    }
                }
                kotlinx.coroutines.delay(POLL_INTERVAL_MS)
            }
        }
    }

    /**
     * Reduce one event. Per-wallet results settle the state
     * (Synced/Error); the pass boundary carries the authoritative timestamp.
     * `PlatformAddressResult` alone doesn't publish `Synced` until we have
     * the pass boundary, matching Swift's per-result-then-snapshot flow.
     */
    internal fun reduce(event: WalletSyncEvent) {
        when (event) {
            is WalletSyncEvent.PlatformAddressResult -> {
                if (!event.success) {
                    _state.value = PlatformSyncState.Error(
                        event.errorMessage ?: "platform-address sync failed",
                    )
                } else {
                    _state.value = PlatformSyncState.Synced(
                        syncHeight = event.newSyncHeight,
                        syncTimestamp = event.newSyncTimestamp,
                    )
                }
            }

            is WalletSyncEvent.PlatformAddressPassCompleted -> {
                // A zero-wallet pass still completed — settle out of Syncing.
                if (event.walletCount == 0 && _state.value is PlatformSyncState.Syncing) {
                    _state.value = PlatformSyncState.Synced(0, event.syncUnixSeconds)
                }
            }

            is WalletSyncEvent.Error -> {
                _state.value = PlatformSyncState.Error(event.message)
            }

            else -> Unit // shielded / generic events are not this service's concern
        }
    }

    /**
     * Request a single user-initiated pass — port of Swift `manualSync()`.
     * Flips to [PlatformSyncState.Syncing] optimistically; the completion
     * event settles it. Forwards to the manager (a request, not
     * orchestration). No-op when unconfigured.
     */
    suspend fun manualSync() {
        val m = manager ?: return
        _state.value = PlatformSyncState.Syncing
        try {
            m.startPlatformAddressSync()
        } catch (e: Exception) {
            _state.value = PlatformSyncState.Error(e.message ?: "manual sync failed")
        }
    }

    /**
     * Clear all synced platform-address data for [network] — port of Swift
     * `PlatformBalanceSyncService.clearLocalState(...)`, the Sync-tab "Clear"
     * action (#3959). **Fail-closed**, in strict order:
     *
     * 1. **Native reset first** — [PlatformWalletManager.resetPlatformAddressSyncState]
     *    quiesces the loop and drops each wallet's balances + the provider
     *    watermark/seed. If it throws, surface the error and clear NOTHING
     *    else (leaving the display and disk intact so the state stays
     *    consistent). When no manager is bound this step is skipped (matching
     *    Swift's `if let walletManager`).
     * 2. **Room clear** — zero the active network's address rows in place
     *    (preserving derivation metadata) and delete the active network's
     *    sync-state watermark; other networks are untouched. If it throws,
     *    surface the error and skip the display clear.
     * 3. **Display clear** — only on full success, reset [state] to
     *    [PlatformSyncState.Idle]; the balance/active-address Room `Flow`s
     *    settle to zero automatically from the disk clear.
     *
     * The loop is deliberately left stopped — data stays cleared until the
     * user explicitly resyncs (via [manualSync] / start).
     *
     * @param network the active network whose synced data to clear.
     * @param walletIdsOnNetwork the 32-byte ids of wallets on [network]
     *   (address rows carry no network column, so they are scoped by this
     *   wallet-id pivot — mirror of Swift's `walletIdsOnNetwork`).
     */
    suspend fun clearLocalState(
        network: org.dashfoundation.dashsdk.Network,
        walletIdsOnNetwork: List<ByteArray>,
    ) {
        // 1. Native reset first (fail-closed).
        val m = manager
        if (m != null) {
            try {
                m.resetPlatformAddressSyncState()
            } catch (e: Exception) {
                _state.value = PlatformSyncState.Error(
                    e.message ?: "failed to reset platform-address sync state",
                )
                return
            }
        }

        // 2. Room clear (fail-closed): zero the active-network address rows in
        //    place + delete the active-network watermark, in one transaction
        //    (mirror of Swift's single `save()`).
        try {
            database.withTransaction {
                database.platformAddressDao().zeroBalancesForWallets(
                    walletIdsOnNetwork,
                    System.currentTimeMillis(),
                )
                database.platformAddressDao().deleteSyncStatesByNetwork(network.ffiValue)
            }
        } catch (e: Exception) {
            _state.value = PlatformSyncState.Error(
                e.message ?: "failed to clear platform-address data",
            )
            return
        }

        // 3. Display clear — only on full success.
        _state.value = PlatformSyncState.Idle
    }

    /**
     * Unbind and reset to [PlatformSyncState.Idle] — port of Swift
     * `reset()`. Cancels subscriptions but keeps the service reusable via a
     * later [configure].
     */
    fun reset() {
        eventJob?.cancel()
        eventJob = null
        pollJob?.cancel()
        pollJob = null
        manager = null
        _state.value = PlatformSyncState.Idle
    }

    /** Tear down the service scope permanently. */
    fun close() {
        scope.cancel()
    }

    private companion object {
        /** Liveness poll cadence — matches the manager's 1 Hz progress poll. */
        const val POLL_INTERVAL_MS = 1_000L
    }
}
